src/metabase/models/dashboard_card.clj

```diff
@@ -135,24 +137,26 @@
                                 [:= :newcard.archived false]
                                 [:= :dashcard.id (:id dashcard)]]}))
 
-(s/defn update-dashboard-card-series!
-  "Update the DashboardCardSeries for a given DashboardCard and return the updated DashboardCardSeries.
-   `card-ids` should be a definitive collection of *all* IDs of cards for the dashboard card in the desired order.
-
-   *  If an ID in `card-ids` has no corresponding existing DashboardCardSeries object, one will be created.
-   *  If an existing DashboardCardSeries has no corresponding ID in `card-ids`, it will be deleted.
-   *  All cards will be updated with a `position` according to their place in the collection of `card-ids`"
-  {:arglists '([dashboard-card card-ids])}
-  [{:keys [id]} :- {:id su/IntGreaterThanZero, s/Keyword s/Any}
-   card-ids     :- [su/IntGreaterThanZero]]
-  ;; first off, just delete all series on the dashboard card (we add them again below)
-  (t2/delete! DashboardCardSeries :dashboardcard_id id)
-  ;; now just insert all of the series that were given to us
-  (when (seq card-ids)
-    (let [cards (map-indexed (fn [i card-id]
-                               {:dashboardcard_id id, :card_id card-id, :position i})
-                             card-ids)]
-      (t2/insert! DashboardCardSeries cards))))
+(defn update-dashboard-cards-series!
+  "Batch update the DashboardCardSeries for multiple DashboardCards.
+  Each `card-ids` list should be a definitive collection of *all* IDs of cards for the dashboard card in the desired order.
+
+  *  If an ID in `card-ids` has no corresponding existing DashboardCardSeries object, one will be created.
+  *  If an existing DashboardCardSeries has no corresponding ID in `card-ids`, it will be deleted.
+  *  All cards will be updated with a `position` according to their place in the collection of `card-ids`"
+  {:arglists '([dashcard-id->card-ids])}
+  [dashcard-id->card-ids]
+  (when-not (empty? dashcard-id->card-ids)
+    ;; first off, just delete all series on the dashboard card (we add them again below)
+    (t2/delete! DashboardCardSeries :dashboardcard_id [:in (keys dashcard-id->card-ids)])
+    ;; now just insert all of the series that were given to us
+    (when-let [card-series (->> (for [[dashcard-id card-ids] dashcard-id->card-ids]
+                                  (map-indexed (fn [i card-id]
+                                                 {:dashboardcard_id dashcard-id, :card_id card-id, :position i})
+                                               card-ids))
+                                flatten
+                                (remove nil?))]
+      (t2/insert! DashboardCardSeries card-series))))
```

Comment:
1. `remove` cannot return a falsy value, so `when-let` is just `let`.
2. I'm not sure how `nil`s can show up in the list at all, we seem to be putting maps there.
3. I think we can get rid of `flatten` by continuing the list comprehension.
```suggestion
    (when-let [card-series (seq (for [[dashcard-id card-ids] dashcard-id->card-ids
                                      [i card-id] (map-indexed vector card-ids)]
                                  {:dashboardcard_id dashcard-id, :card_id card-id, :position i}))]
      (t2/insert! DashboardCardSeries card-series))))
```

Reply:
This is much better and your assessment is correct. Thanks

Reply:
0d7da010b2